python/ray/dag/dag_operation_future.py

```diff
@@ -0,0 +1,88 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Generic, List, Optional, TypeVar, Union
+
+T = TypeVar("T")
+
+
+class DAGOperationFuture(ABC, Generic[T]):
+    @abstractmethod
+    def wait(self, waiter: Optional[Any] = None):
+        raise NotImplementedError
+
+
+class ReadyFuture(DAGOperationFuture):
+    def __init__(self, result):
+        self._result = result
+
+    def wait(self, waiter: Optional[Any] = None):
+        return self._result
+
+
+class ListFuture(DAGOperationFuture):
+    def __init__(self, futures: List[DAGOperationFuture]):
+        self._futures = futures
+
+    def wait(self, waiter: Optional[Any] = None):
+        results = []
+        for future in self._futures:
+            results.append(future.wait(waiter))
+        return results
+
+
+class WrappedFuture(DAGOperationFuture):
```

Comment:
Why do we need this?

Reply:
Now that we use futures everywhere, whenever we want to examine the value, it may not be appropriate to wait right away, which changes the operation (including _GPUFuture) from async to sync.

WrappedFuture and ListFuture are basically introduced to support delayed executions: delay any action until the future is waited upon. WrappedFuture wraps a lambda, and ListFuture wraps extract-from-list operation.

Reply:
Removed.